safety_prototype/opentelemetry-api/src/opentelemetry/_safety.py

```diff
@@ -0,0 +1,50 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import wraps
+from warnings import warn, resetwarnings
+from traceback import format_exception
+from sys import exc_info
+
+
+def _safe_function(predefined_return_value):
+    """
+    This is the safety mechanism mentioned in the README file.
+
+    This is used as a decorator on every function or method in the API.
+    """
+
+
+    def internal(function):
+        @wraps(function)
+        def wrapper(*args, **kwargs):
+
+            # This is the try / except block mentioned in the README file.
+            try:
+                exception = None
+                return function(*args, **kwargs)
+            except Exception:  # pylint: disable=broad-except
+                exception = "".join(format_exception(*exc_info()))
+
+            if exception is not None:
```

Comment:
Hm, I think there may be a misunderstanding here. The mechanism proposed here is intended to guard against exceptions raised in the SDK. Any exception raised in the application code directly that is caused by the calling of a function or method outside of the OpenTelemetry API is not protected and will be raised normally.

So, you mention "user made exceptions". If you mean by that exceptions raised by code not called by functions or methods in the OpenTelemetry API, then this mechanism will not cause any issues, these "user made exceptions" will be raised normally.

If with "user made exceptions" you mean exceptions in any part of the SDK that were "intentionally coded", for example something like this:

@_safe_function(X)
def some_function(a, b):
    ...
    if some_condition:
        raise Exception("Some exception")
    ...

then the mechanism will also catch these exceptions. This is intentional. The spec says that any exception of this kind must be handled by the safety mechanism, regardless of how "intentional" it is or not.

Reply:
No, What I meant is something where the OpenTelemetry wrapper is "safe", and the original function which it wraps raises an exception (For example, `raise Http404()` in Django). In this case, the original function will raise the exception, which will raise to the OpenTelemetry Django wrapper, which will raise to the safety wrapper. Perhaps Django isn't the best example as it uses a middleware object and doesn't directly call `wrapt.wrap_function_wrapper`, but the SDK does wrap user-defined functions.

Reply:
Ok, sorry but I don't understand what's wrong here. It is intentional to guard against any kind of exception raised in the scope of the safety mechanism, regardless of who raised it or whatever intention was behind raising it. This is a specification requirement, because if we let an exception to be raised it will crash the application which must not happen. So, if your concern is that this mechanism may end up "swallowing" an user made exception, then yes, this mechanism will do that and it is intended to do so.

Reply:
Alright. I just wanted to bring this up because it alters the behavior, so it will a known side effect

Reply:
I think this is a lot more nuanced than that. I agree with Diego about swallowing exceptions from user defined code but I think the "user defined code" needs a definition in this context. For example, we should definitely swallow exceptions raised by user provider callbacks like pre/post request instrumentation hooks. However, I don't think we should swallow exceptions raised by user/library code that might be instrumented. Django404 is actually a good example of this. Instrumentation (manual or auto) must not alter the behavior of the program and if we start swallowing all exceptions raised by code executed inside a span context (think `with start_span()`) we'll be altering behavior of programs. I think this is what @oxeye-nikolay was thinking of.

Reply:
Ah, ok, I see, but I have a question here, what is the expected behavior when Django404 exception happens?

Does this exception end up being raised up until it crashes the application? Or is it handled somehow differently by another mechanism that does not cause the application to crash?

Reply:
It is handled by the Django Framework so the developer can decide how the 404 response looks. If we do choose to address this issue because it might alter the behavior of the program, it can probably be done by parsing the trace stack, but I'm not sure it's something that we want to do because this logic might get complicated.
What do you think about it?

Reply:
I think we can revisit this when we go over all instrumentation to ensure we don't raise errors there. We shouldn't do anything special to handle any known exceptions differently. 

We just need to ensure that our context managers (`start_as_current_span`, `use_span`) don't swallow exceptions raised from within the context. May be this means they shouldn't be decorated and error protection should be coded manually into their bodies or may be they can catch all exceptions but annotate the ones coming from user code so the decorator can re-raise them but that sounds unnecessarily complicated.

Reply:
> but annotate the ones coming from user code so the decorator can re-raise them...

Hm, but does this mean that the re-raised exceptions will be able to crash the application?

Reply:
I think so, yes. If the application would crash without instrumentation then it should crash with instrumentation as well, right? We should ensure that:

1. SDK/instrumentation code never raises exceptions.
2. For instrumented code paths, we record exceptions as events if the user wants to and then re-raise.

This is how `use_span` works today: https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-api/src/opentelemetry/trace/__init__.py#L530-L559

As an example, if I have a function that looks like:

```python
def my_func():
    raise Exception()
```

and I instrument it as:

```python
def my_func():
    with tracer.start_as_current_span("span1"):
        raise Exception()
```

then the only change that should happen is that now my program should export a span called `span1` with an event which has information about the exception I'm raising. It should not swallow the exception. However, if an exception is raised inside `start_as_current_span()` function or any other function it calls directly, that exception should be swallowed and never reach my code.

Reply:
Ah, I see.

Ok, I think the current design works as you want it to work, because it will only catch exceptions raised in the *execution* of `start_as_current_span` (not in its `with`-managed context) in the example above.